Automattic/woocommerce-payments
# Enhance plugin logging to provide more context information and request IDs

Fixes #9892 

#### Changes proposed in this Pull Request

 - Include unique request ID, generated upon plugin initialization, into each log entry.
 - Include sequential entry numbers into the log.
 - Use JSON to format structured data in API client.
 - Ability to provide additional context for the request entry, e.g. order ID, payment intention ID, which can be used later for parsing.
 - Improve formatting of multi-line log entries to have the same prefix as the first line for better readability and processing by tools like `grc`.
 - Request and Response in the API client are paired with unique IDs to make it easier to match Response to a Request, in case multiple requests have happened.

#### Testing instructions

 * Automated tests shall pass.
 * Checkout branch and visit admin pages of the plugin. Check the `woopayments-YYYY-MM-DD...` logs in the `wp-content/uploads/wc-logs/` and see that the formatting is in place, e.g. like below:
 
```
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004 REQUEST GET https://public-api.wordpress.com/wpcom/v2/sites/%s/wcpay/disputes?test_mode=1&pagesize=50&search%5B0%5D=warning_needs_response&search%5B1%5D=needs_response
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004 CONTEXT: {
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004     "WP_User": "Guest (non logged-in user)",
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004     "HTTP_REFERER": "http:\/\/wcpay-debian.vm.local:8082\/wp-admin\/admin.php?page=wc-admin&path=paymentstransactions",
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004     "HTTP_USER_AGENT": "Mozilla\/5.0 (Macintosh; Intel Mac OS X 10.15; rv:133.0) Gecko\/20100101 Firefox\/133.0",
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004     "REQUEST_URI": "\/wp-admin\/admin.php?page=wc-admin&path=paymentstransactionsdetails&id=pi_3QRxAsR5iZ61kaPg13Bz8Ph6&transaction_id=txn_3QRxAsR5iZ61kaPg1mZ5lQYe&transaction_type=charge",
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004     "DOING_AJAX": "",
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004     "DOING_CRON": "",
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004     "WP_CLI": ""
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0004 }
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0005 HEADERS: {
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0005     "Content-Type": "application\/json; charset=utf-8",
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0005     "User-Agent": "WooCommerce Payments\/8.5.1"
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0005 }
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0006 RESPONSE: {
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0006     "data": []
2024-12-05T08:38:43+00:00 INFO 6751668fafe68-0006 }
```

or with `grc` running:

<img width="1920" alt="image" src="https://github.com/user-attachments/assets/ddd01574-ea99-467e-ac43-b4408aa8a6bb">

-------------------


**Post merge**
